src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,43 +1,61 @@
-        result.add(new Product(8L, "xperiaxz", "Sony Xperia XZ", new BigDecimal(120), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Sony/Sony%20Xperia%20XZ.jpg"));
-        result.add(new Product(9L, "nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
-        result.add(new Product(10L, "palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
-        result.add(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
-        result.add(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
-        result.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
     }
+    
+    public static ProductDao getInstance(){
+        if(instance == null){
+            instance = new ArrayListProductDao();
+        }
+        return  instance;
+    }
+    
+    public Comparator<Product> getComparator(SortField field, SortPrice price){
+        if(field.equals(SortField.no) && price.equals(SortPrice.no)){
+            return Comparator.comparing(Product::getStock);
+        }
+        if(field.equals(SortField.asc)){
+            return Comparator.comparing(Product::getDescription);
+        }
+        if(field.equals(SortField.desc)){
+            return Comparator.comparing(Product::getDescription).reversed();
+        }
+        
+        if(price.equals(SortPrice.asc)){
+            return Comparator.comparing(Product::getPrice);
+        }
+        else{
+            return Comparator.comparing(Product::getPrice).reversed();
+        }
+    }
+    
     @Override
     public Product getProduct(Long id) {
         return result.stream().
-                filter(x-> x.getId().equals(id)).
+                filter(x-> x.getId()!=null && x.getId().equals(id)).
                 findFirst().
                 orElseThrow(IllegalArgumentException::new);
     }
     
     @Override
-    public List<Product> findProducts() {
-        return result.stream().
+    public List<Product> findProducts(String query, SortField field, SortPrice price) {
+            return result.stream().
                 filter(x-> x.getPrice()!=null && x.getStock()>0).
+                filter(x-> query==null || x.getDescription().contains(query)).
```

Comment:
More functionality expected here: the task is "The search must work using OR clause. The example above means: search all products having description containing “Samsung” or “S” or “II”. Products having all search terms must come first in the search result."